### ISSUE TITLE
How to make the Checkpoint store moments and other relevant variables in tf.train Optimizers.

### ISSUE THREAD

**Issue by yaoyiran** [tensorflow/tensorflow#30086]
I encountered a problem when my code stopped for some reason on my machine, so I had to restart my code and continue the training process by loading the latest checkpoint file. I found that the performance is not consistent before and after the checkpoint that I loaded and the performance dropped a lot. So, since my code uses tf.train.AdamOptimizer, I guess that the checkpoint doesn't store the moment vectors and the gradients in the previous steps, and when I load the checkpoint the moment vectors are initialized as zeros. Am I correct? Is there any method that can help store relevant vectors for the Adamopotimizer in the checkpoints so that if my machine is down again, restarting from the latest checkpoint will not influence anything? Thanks!

**Comment by allenlavoie** [comment]
Which APIs are you using to save and restore? `tf.train.Saver` and `tf.train.Checkpoint` will both save optimizer slot variables ([the guide has a diagram including slot variables](https://github.com/tensorflow/docs/blob/master/site/en/r2/guide/checkpoints.ipynb) under "Loading Mechanics").

**Comment by yaoyiran** [comment]
@allenlavoie Thanks for your reply! I am using saver = tf.train.Saver() API and use saver.save(), saver.restore() to save and restore models. But could you explain why the performance obviously dropped when I restored the model from the latest checkpoint and went for more epoches? The "latest checkpoint" that I restored is just like a watershed in terms of performance.

**Comment by ravikyram** [comment]
Can you please help us with Tensorflow version and full code snippet to reproduce the issue.Thanks.

**Comment by allenlavoie** [comment]
One common issue is creating the Saver too early; by default it picks up variables from the global collection, so if the Optimizer's minimize call hasn't run yet then it won't pick up those variables.
